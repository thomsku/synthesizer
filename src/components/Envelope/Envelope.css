.graph {
  height: 10em;
  width: 25em;
  margin-top: 0.5em;
}

.slider {
  padding: 22px 0px;
}