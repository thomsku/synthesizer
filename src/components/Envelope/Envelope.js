import React from "react";
import EnvelopeGraph from "adsr-envelope-graph";
import Slider from "@material-ui/lab/Slider";
import Typography from "@material-ui/core/Typography";
import FormLabel from "@material-ui/core/FormLabel";
import {withStyles} from "@material-ui/core/styles";
import {fade} from "@material-ui/core/styles/colorManipulator";
import PropTypes from "prop-types";
import "./Envelope.css";

const StyledSlider = withStyles({
  thumb: {
    height: 24,
    width: 24,
    backgroundColor: "#fff",
    "&$focused, &:hover": {
      boxShadow: `0px 0px 0px ${8}px ${fade("#de235b", 0.16)}`,
    },
    "&$activated": {
      boxShadow: `0px 0px 0px ${8 * 1.5}px ${fade("#de235b", 0.16)}`,
    },
    "&$jumped": {
      boxShadow: `0px 0px 0px ${8 * 1.5}px ${fade("#de235b", 0.16)}`,
    },
  },
  track: {
    backgroundColor: "#fff",
    height: 8,
  },
  trackAfter: {
    backgroundColor: "#fff",
  },
  focused: {},
  activated: {},
  jumped: {},
})(Slider);

class Envelope extends React.Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event, value, type) {
    this.props.setEnvelope(type, value);
  }
  render() {
    return (
      <div className="envelope">
        <FormLabel component="legend">Envelope</FormLabel>
        <div className="graph">
          <EnvelopeGraph style={
            {height: "100%", width: "100%"}}
          a={this.props.envelope.attack}
          d={this.props.envelope.decay}
          s={this.props.envelope.sustain}
          r={this.props.envelope.release}
          />
        </div>
        <div className="sliders">

          <Typography id="label">A</Typography>
          <StyledSlider
            id="attack"
            className="slider"
            value={this.props.envelope.attack}
            min={0.01}
            max={2}
            aria-labelledby="label"
            onChange={(event, value) => {
              this.handleChange(event, value, "attack");
            }}
          />
          <Typography id="label">D</Typography>
          <StyledSlider
            id="attack"
            className="slider"
            value={this.props.envelope.decay}
            min={0.01}
            max={2}
            aria-labelledby="label"
            onChange={(event, value) => {
              this.handleChange(event, value, "decay");
            }}
          />
          <Typography id="label">S</Typography>
          <StyledSlider
            id="attack"
            className="slider"
            value={this.props.envelope.sustain}
            min={0}
            max={1}
            aria-labelledby="label"
            onChange={(event, value) => {
              this.handleChange(event, value, "sustain");
            }}
          />
          <Typography id="label">R</Typography>
          <StyledSlider
            id="attack"
            className="slider release"
            value={this.props.envelope.release}
            min={0.01}
            max={2}
            aria-labelledby="label"
            onChange={(event, value) => {
              this.handleChange(event, value, "release");
            }}
          />
        </div>
      </div>
    );
  }
}

Envelope.propTypes = {
  setEnvelope: PropTypes.func.isRequired,
  envelope: PropTypes.objectOf(PropTypes.number).isRequired,
};

export default Envelope;
