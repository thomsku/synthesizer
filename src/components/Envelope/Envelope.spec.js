import React from "react";
import {shallow} from "enzyme";
import Envelope from "./Envelope";

describe("Envelope", () => {
  let wrapper;
  beforeEach(() => wrapper = shallow(<Envelope setEnvelope={jest.fn()} envelope={{attack: 0.1, decay: 0.01, sustain: 1, release: 0.1}}/>));

  it("should render correctly", () => expect(wrapper).toMatchSnapshot());

  it("should render 3 <div />", () => {
    expect(wrapper.find("div")).toHaveLength(3);
  });
});
