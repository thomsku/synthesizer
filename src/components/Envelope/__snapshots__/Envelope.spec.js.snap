// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Envelope should render correctly 1`] = `
<div
  className="envelope"
>
  <WithStyles(WithFormControlContext(ForwardRef(FormLabel)))
    component="legend"
  >
    Envelope
  </WithStyles(WithFormControlContext(ForwardRef(FormLabel)))>
  <div
    className="graph"
  >
    <EnvelopeGraph
      a={0.1}
      d={0.01}
      r={0.1}
      s={1}
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </div>
  <div
    className="sliders"
  >
    <WithStyles(ForwardRef(Typography))
      id="label"
    >
      A
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(WithStyles(ForwardRef(Slider)))
      aria-labelledby="label"
      className="slider"
      id="attack"
      max={2}
      min={0.01}
      onChange={[Function]}
      value={0.1}
    />
    <WithStyles(ForwardRef(Typography))
      id="label"
    >
      D
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(WithStyles(ForwardRef(Slider)))
      aria-labelledby="label"
      className="slider"
      id="attack"
      max={2}
      min={0.01}
      onChange={[Function]}
      value={0.01}
    />
    <WithStyles(ForwardRef(Typography))
      id="label"
    >
      S
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(WithStyles(ForwardRef(Slider)))
      aria-labelledby="label"
      className="slider"
      id="attack"
      max={1}
      min={0}
      onChange={[Function]}
      value={1}
    />
    <WithStyles(ForwardRef(Typography))
      id="label"
    >
      R
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(WithStyles(ForwardRef(Slider)))
      aria-labelledby="label"
      className="slider release"
      id="attack"
      max={2}
      min={0.01}
      onChange={[Function]}
      value={0.1}
    />
  </div>
</div>
`;
